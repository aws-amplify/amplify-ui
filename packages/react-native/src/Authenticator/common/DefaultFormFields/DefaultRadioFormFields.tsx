import React from 'react';
import { censorAllButFirstAndLast, censorPhoneNumber } from '@aws-amplify/ui';

import { Radio, RadioGroup } from '../../../primitives';
import { DefaultRadioFormFieldsProps } from './types';

const censorContactInformation = (name: string, value: string): string => {
  let censoredVal = value;
  if (name === 'email') {
    const splitEmail = value.split('@');
    const censoredName = censorAllButFirstAndLast(splitEmail[0]);

    censoredVal = `${censoredName}@${splitEmail[1]}`;
  } else if (name === 'phone_number') {
    censoredVal = censorPhoneNumber(value);
  }
  return censoredVal;
};

const DefaultRadioFormFields = ({
  fields,
  fieldContainerStyle,
  fieldLabelStyle,
  isPending,
  style,
}: DefaultRadioFormFieldsProps): JSX.Element => {
  return (
    <RadioGroup disabled={isPending} style={style}>
      {(fields ?? []).map(({ name, value, ...props }) => (
        <Radio
          {...props}
          key={value}
          // value has to be name, because Auth is only interested in the
          // string "email" or "phone_number", not the actual value
          value={name}
          label={censorContactInformation(name, value)}
          labelStyle={fieldLabelStyle}
          style={fieldContainerStyle}
        />
      ))}
    </RadioGroup>
  );
};

DefaultRadioFormFields.displayName = 'FormFields';

export default DefaultRadioFormFields;
