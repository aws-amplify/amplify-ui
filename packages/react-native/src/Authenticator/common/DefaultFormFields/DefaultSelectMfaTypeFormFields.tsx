import React from 'react';

import { DefaultRadioFormFieldsProps } from './types';

import { RadioGroup } from '../../../primitives/RadioGroup';
import { Radio } from '../../../primitives/Radio';

const DefaultSelectMfaTypeFormFields = ({
  fields = [],
  fieldContainerStyle,
  fieldLabelStyle,
  isPending,
  style,
}: DefaultRadioFormFieldsProps): React.JSX.Element => {
  // set initial value for radio field based on selected bool
  const initialValue = fields.find((field) => !!field.selected)?.value;
  return (
    <RadioGroup disabled={isPending} style={style} initialValue={initialValue}>
      {fields.map(({ value, label, ...props }) => (
        <Radio
          {...props}
          key={value}
          value={value}
          label={label}
          labelStyle={fieldLabelStyle}
          style={fieldContainerStyle}
        />
      ))}
    </RadioGroup>
  );
};

DefaultSelectMfaTypeFormFields.displayName = 'FormFields';

export default DefaultSelectMfaTypeFormFields;
