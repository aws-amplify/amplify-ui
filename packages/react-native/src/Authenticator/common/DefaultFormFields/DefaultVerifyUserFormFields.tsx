import React from 'react';
import { censorContactMethod, ContactMethod } from '@aws-amplify/ui';

import { Radio, RadioGroup } from '../../../primitives';
import { DefaultRadioFormFieldsProps } from './types';

interface AttributeMap {
  email: ContactMethod;
  phone_number: ContactMethod;
}

const attributeMap: AttributeMap = {
  email: 'Email',
  phone_number: 'Phone Number',
};

const DefaultVerifyUserFormFields = ({
  fields = [],
  fieldContainerStyle,
  fieldLabelStyle,
  isPending,
  style,
}: DefaultRadioFormFieldsProps): React.JSX.Element => {
  // set initial value for radio field based on selected bool
  const initialValue = fields.find((field) => !!field.selected)?.name;
  return (
    <RadioGroup disabled={isPending} style={style} initialValue={initialValue}>
      {fields.map(({ name, value, ...props }) => {
        const attributeType = attributeMap[name as keyof AttributeMap];
        return (
          <Radio
            {...props}
            key={value}
            // value has to be name, because Auth is only interested in the
            // string "email" or "phone_number", not the actual value
            value={name}
            label={censorContactMethod(attributeType, value)}
            labelStyle={fieldLabelStyle}
            style={fieldContainerStyle}
          />
        );
      })}
    </RadioGroup>
  );
};

DefaultVerifyUserFormFields.displayName = 'FormFields';

export default DefaultVerifyUserFormFields;
