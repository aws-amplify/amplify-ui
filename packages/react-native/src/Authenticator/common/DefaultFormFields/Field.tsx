import React from 'react';
import { TextInput, TextInputProps } from 'react-native';

import {
  PasswordField,
  PhoneNumberField,
  TextField,
} from '../../../primitives';
import { platform } from '../../../utils';
import { FieldProps } from './types';

const { IS_IOS } = platform;

// to prevent issues with iOS when multiple `TextInput` components have `secureTextEntry`
// set to `true`, insert a "hidden" `TextInput` after each `PasswordField`
// Issue reference: https://github.com/facebook/react-native/issues/21911
const HIDDEN_INPUT_PROPS: TextInputProps = {
  // prevent iOS screen reader from picking up element
  accessibilityElementsHidden: true,
  // prevent `TextInput` from capturing touch events
  pointerEvents: 'none',
  // this workaround requires the `height` and `width` applied to the `TextInput`
  // are greater than `0`
  // NOTE: do not attempt to set an opacity value here to further hide the element,
  // it will cause the issues mitigated by this workaround to re-surface
  style: { backgroundColor: 'transparent', height: 0.1, width: 0.1 },
};

const HiddenInput = () => <TextInput {...HIDDEN_INPUT_PROPS} />;

const Field = ({ type, ...rest }: FieldProps): JSX.Element => {
  const isPassword = type === 'password';
  const Field = isPassword
    ? PasswordField
    : type === 'phone'
    ? PhoneNumberField
    : TextField;

  return IS_IOS && isPassword ? (
    <>
      <Field {...rest} />
      <HiddenInput />
    </>
  ) : (
    <Field {...rest} />
  );
};

Field.displayName = 'Field';
export default Field;
