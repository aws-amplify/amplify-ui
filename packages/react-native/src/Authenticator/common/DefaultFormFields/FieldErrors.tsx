import React from 'react';
import { Text } from 'react-native';

import { FieldErrorsProps } from './types';

export const FieldErrors = ({
  errors,
  style,
}: FieldErrorsProps): JSX.Element | null => {
  if (!(errors?.length > 0)) {
    return null;
  }

  return (
    <>
      {errors.map((error) => (
        <Text key={error} style={style}>
          {error}
        </Text>
      ))}
    </>
  );
};
