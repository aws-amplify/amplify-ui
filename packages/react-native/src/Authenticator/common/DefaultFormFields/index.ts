export { default as DefaultRadioFormFields } from './DefaultRadioFormFields';
export { default as DefaultTextFormFields } from './DefaultTextFormFields';
export { DefaultFormFieldsComponent, DefaultFormFieldsStyle } from './types';
