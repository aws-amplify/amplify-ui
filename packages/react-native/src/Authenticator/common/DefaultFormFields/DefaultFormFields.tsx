import React, { Fragment } from 'react';
import { View } from 'react-native';

import { getErrors } from '@aws-amplify/ui';
import { DefaultComponents, TextFieldOptionsType } from '../../Defaults/types';
import { FieldErrors } from './FieldErrors';
import { styles } from './styles';
import {
  PasswordField,
  PhoneNumberField,
  TextField,
} from '../../../primitives';

const DefaultFormFields: DefaultComponents<TextFieldOptionsType>[keyof DefaultComponents]['FormFields'] =
  ({ fields, isPending, validationErrors }) => {
    const fieldStyle = styles.field;
    const formFields = fields.map(({ name, type, ...field }) => {
      const errors = validationErrors
        ? getErrors(validationErrors?.[name])
        : [];

      const hasError = errors?.length > 0;

      const Field =
        type === 'password'
          ? PasswordField
          : type === 'phone'
          ? PhoneNumberField
          : TextField;

      return (
        <Fragment key={name}>
          <Field
            {...field}
            autoCapitalize="none"
            // disabled style is applying to Label
            disabled={isPending}
            error={hasError}
            key={name}
            style={fieldStyle}
          />
          <FieldErrors errors={errors} style={styles.error} />
        </Fragment>
      );
    });
    return <View>{formFields}</View>;
  };

DefaultFormFields.displayName = 'FormFields';

export default DefaultFormFields;
