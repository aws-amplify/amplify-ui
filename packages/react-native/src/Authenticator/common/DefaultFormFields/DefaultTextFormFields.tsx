import React, { Fragment } from 'react';
import { View } from 'react-native';

import { getErrors } from '@aws-amplify/ui';

import Field from './Field';
import { FieldErrors } from './FieldErrors';
import {
  // DefaultTextFormFieldsComponent,
  DefaultTextFormFieldsProps,
} from './types';

const DefaultTextFormFields = ({
  fieldContainerStyle,
  fieldErrorsContainer,
  fieldErrorStyle,
  fieldStyle,
  fields,
  isPending = false,
  style,
  validationErrors,
}: DefaultTextFormFieldsProps): JSX.Element => {
  const formFields = (fields ?? []).map(({ name, type, ...field }) => {
    const errors = validationErrors ? getErrors(validationErrors?.[name]) : [];

    const hasError = errors?.length > 0;

    return (
      <Fragment key={name}>
        <Field
          {...field}
          disabled={isPending}
          error={hasError}
          fieldStyle={fieldStyle}
          key={name}
          style={fieldContainerStyle}
          type={type}
        />
        <FieldErrors
          errors={errors}
          errorStyle={fieldErrorStyle}
          style={fieldErrorsContainer}
        />
      </Fragment>
    );
  });
  return <View style={style}>{formFields}</View>;
};

DefaultTextFormFields.displayName = 'FormFields';

export default DefaultTextFormFields;
