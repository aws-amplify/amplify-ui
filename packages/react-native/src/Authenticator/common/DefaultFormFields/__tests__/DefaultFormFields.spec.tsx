import React from 'react';
import { render } from '@testing-library/react-native';

import DefaultFormFields from '../DefaultTextFormFields';

describe('DefaultFormFields', () => {
  it('renders as expected', () => {
    const { toJSON } = render(
      <DefaultFormFields fields={[]} isPending={false} />
    );
    expect(toJSON()).toMatchSnapshot();

    expect(toJSON()).toBeDefined();
  });
});
