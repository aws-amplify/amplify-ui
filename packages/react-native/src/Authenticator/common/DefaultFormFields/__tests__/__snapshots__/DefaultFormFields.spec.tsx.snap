// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DefaultFormFields renders as expected 1`] = `<View />`;
